edx_sga/sga.py

```diff
@@ -0,0 +1,530 @@
+        answer = self.get_submission()['answer']
+        path = self._file_storage_path(answer['sha1'], answer['filename'])
+        return self.download(path, answer['mimetype'], answer['filename'])
+
+    @XBlock.handler
+    def download_annotated(self, request, suffix=''):
+        path = self._file_storage_path(
+            self.annotated_sha1,
+            self.annotated_filename,
+        )
+        return self.download(
+            path,
+            self.annotated_mimetype,
+            self.annotated_filename
+        )
+
+    @XBlock.handler
+    def staff_download(self, request, suffix=''):
+        require(self.is_course_staff())
+        submission = self.get_submission(request.params['student_id'])
+        answer = submission['answer']
+        path = self._file_storage_path(answer['sha1'], answer['filename'])
+        return self.download(path, answer['mimetype'], answer['filename'])
+
+    @XBlock.handler
+    def staff_download_annotated(self, request, suffix=''):
+        require(self.is_course_staff())
+        module = StudentModule.objects.get(pk=request.params['module_id'])
+        state = json.loads(module.state)
+        path = self._file_storage_path(
+            state['annotated_sha1'],
+            state['annotated_filename']
+        )
+        return self.download(
+            path,
+            state['annotated_mimetype'],
+            state['annotated_filename']
+        )
+
+    def download(self, path, mimetype, filename):
+        BLOCK_SIZE = (1 << 10) * 8  # 8kb
+        file = default_storage.open(path)
+        app_iter = iter(partial(file.read, BLOCK_SIZE), '')
+        return Response(
+            app_iter=app_iter,
+            content_type=mimetype,
+            content_disposition="attachment; filename=" + filename)
+
+    @XBlock.handler
+    def get_staff_grading_data(self, request, suffix=''):
+        require(self.is_course_staff())
+        return Response(json_body=self.staff_grading_data())
+
+    @XBlock.handler
+    def enter_grade(self, request, suffix=''):
+        require(self.is_course_staff())
+        module = StudentModule.objects.get(pk=request.params['module_id'])
+        state = json.loads(module.state)
+        score = int(request.params['grade'])
+        if self.is_instructor():
+            uuid = request.params['submission_id']
+            submissions_api.set_score(uuid, score, self.max_score())
+        else:
+            state['staff_score'] = score
+        state['comment'] = request.params.get('comment', '')
+        module.state = json.dumps(state)
+        module.save()
+
+        return Response(json_body=self.staff_grading_data())
+
+    @XBlock.handler
+    def remove_grade(self, request, suffix=''):
+        require(self.is_course_staff())
+        student_id = request.params['student_id']
+        submissions_api.reset_score(student_id, self.course_id, self.block_id)
+        module = StudentModule.objects.get(pk=request.params['module_id'])
```

Comment:
For scores and submissions, it would be ideal to preserve all data for analytics. The submissions API will by default return the latest score, so adding a new score of 0 should clear the current grade. 

Resetting the submission is another scenario. I believe you can just clear the state from the instructor dashboard, for a quick solution, which should accomplish what is being done here.